### PR TITLE
Artifact enable

### PR DESCRIPTION
**Pull request by yesnault**

### CODE REVIEW

File: engine/api/action.go [hunk 1]
```diff
@@ -195,16 +195,15 @@ func updateActionHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *
 	defer tx.Rollback()
 
 	a.ID = actionDB.ID
-	err = action.UpdateActionDB(tx, a, c.User.ID)
-	if err != nil {
+
+	if err = action.UpdateActionDB(tx, a, c.User.ID); err != nil {
 		log.Warning("updateAction: Cannot update action: %s\n", err)
 		w.WriteHeader(http.StatusInternalServerError)
```
Comment (sguiheux) [inline engine/api/action.go]: use WritterError

File: engine/api/action.go [hunk 2]
```diff
@@ -246,15 +245,13 @@ func addActionHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *con
 	defer tx.Rollback()
 
 	a.Type = sdk.DefaultAction
-	err = action.InsertAction(tx, a, true)
-	if err != nil {
+	if err = action.InsertAction(tx, a, true); err != nil {
 		log.Warning("Action: Cannot insert action: %s\n", err)
 		w.WriteHeader(http.StatusInternalServerError)
```
Comment (sguiheux) [inline engine/api/action.go]: use WritterError

File: engine/api/action.go [hunk 3]
```diff
@@ -312,25 +310,35 @@ func loadActionHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *co
...
-		a, err = sdk.NewActionFromScript(btes)
-		if err != nil {
+		if a, err = sdk.NewActionFromScript(btes); err != nil {
 			WriteError(w, r, err)
 			return
 		}
+	} else { // a jsonified action is posted in body
+		btes, errRead := ioutil.ReadAll(r.Body)
+		if errRead != nil {
+			w.WriteHeader(http.StatusBadRequest)
```
Comment (sguiheux) [inline engine/api/action.go]: use WritterError

File: engine/api/action.go [hunk 4]
```diff
@@ -312,25 +310,35 @@ func loadActionHandler(w http.ResponseWriter, r *http.Request, db *sql.DB, c *co
...
+			return
+		}
+
+		if a, err = sdk.NewAction("").FromJSON(btes); err != nil {
+			w.WriteHeader(http.StatusBadRequest)
```
Comment (sguiheux) [inline engine/api/action.go]: use WritterError

File: engine/api/action/action.go [hunk 5]
```diff
@@ -419,8 +397,7 @@ func insertAudit(db database.QueryExecuter, actionID, userID int64, change strin
 
 	query := `INSERT INTO action_audit (action_id, user_id, change, versionned, action_json)
 			VALUES ($1, $2, $3, NOW(), $4)`
-	_, err = db.Exec(query, actionID, userID, change, a.JSON())
-	if err != nil {
+	if _, err = db.Exec(query, actionID, userID, change, a.JSON()); err != nil {
```
Comment (sguiheux) [inline engine/api/action/action.go]: err :=

File: sdk/action.go [hunk 6]
```diff
@@ -365,6 +361,29 @@ func UpdateJoinedAction(projectKey, pipelineName string, stage int, a Action) er
...
+	btes, errMarshall := json.Marshal(action)
+	if errMarshall != nil {
+		return nil, errMarshall
+	}
+	data, code, err := Request("POST", path, btes)
+	if err != nil {
+		return nil, err
+	}
+	if code >= 300 {
+		return nil, fmt.Errorf("HTTP %d", code)
```
Comment (fsamin) [inline sdk/action.go]: dommage de ne pas renvoyer l'erreur

File: sdk/action_script_test.go [hunk 7]
```diff
@@ -71,7 +71,7 @@ EOF
 	a, err := NewActionFromScript(b)
 	assert.NotNil(t, a)
 	assert.NoError(t, err)
-	t.Logf("Action : %s", a)
+	//t.Logf("Action : %s", a)
```
Comment (fsamin) [inline sdk/action_script_test.go]: à supprimer ou à garder

File: sdk/action_script_test.go [hunk 8]
```diff
@@ -181,3 +181,115 @@ steps  = [{
...
+
+		`)
+
+	a, err := NewActionFromScript(b)
+	assert.NotNil(t, a)
+	assert.NoError(t, err)
+	t.Logf("Action : %v", a)
+
+	assert.Equal(t, ScriptAction, a.Actions[0].Name)
+	//assert.Equal(t, BuiltinAction, a.Actions[0].Type)
```
Comment (fsamin) [inline sdk/action_script_test.go]: decommenter les assertions

File: sdk/cli/cds/action/add.go [hunk 9]
```diff
@@ -2,39 +2,52 @@ package action
 
 import (
 	"fmt"
+	"io/ioutil"
 	"strings"
 
 	"github.com/ovh/cds/sdk"
-
 	"github.com/spf13/cobra"
 )
```
Comment (fsamin) [inline sdk/cli/cds/action/add.go]: réorganiser les imports

File: sdk/cli/cds/action/add.go [hunk 10]
```diff
@@ -44,20 +57,55 @@ func addAction(cmd *cobra.Command, args []string) {
...
+		parameters = append(parameters, p)
+	}
+	return parameters
+}
+
+// addActionFromScript adds an action from an URL or a local file
+func addActionFromScript() error {
+	if strings.HasPrefix(cmdActionAddParams.URL, "http") {
+		_, err := sdk.NewActionFromRemoteScript(cmdActionAddParams.URL, getCmdParameters())
+		return err
```
Comment (fsamin) [inline sdk/cli/cds/action/add.go]: if err != nil

File: sdk/cli/cds/action/add.go [hunk 11]
```diff
@@ -44,20 +57,55 @@ func addAction(cmd *cobra.Command, args []string) {
...
+			Name: p,
+			Type: sdk.StringParameter,
+		}
+		parameters = append(parameters, p)
+	}
+	return parameters
+}
+
+// addActionFromScript adds an action from an URL or a local file
+func addActionFromScript() error {
```
Comment (fsamin) [inline sdk/cli/cds/action/add.go]: return *sdk.Action, error

File: engine/api/worker/model.go [hunk 12]
```diff
@@ -24,7 +24,7 @@ func InsertWorkerModel(db *sql.DB, model *sdk.Model) error {
 func LoadWorkerModels(db database.Querier) ([]sdk.Model, error) {
 	query := `SELECT worker_model.id, worker_model.type, worker_model.name, worker_model.image, worker_model.owner_id , "user".username
 	          FROM worker_model
-	          JOIN "user" ON "user".id =  worker_model.owner_id
+	          LEFT JOIN "user" ON "user".id =  worker_model.owner_id
```
Comment (proullon) [inline engine/api/worker/model.go]: Why would you have a model without owner ?

File: sdk/action.go [hunk 13]
```diff
@@ -365,6 +361,34 @@ func UpdateJoinedAction(projectKey, pipelineName string, stage int, a Action) er
 	return nil
 }
 
+//LoadAction creates an action on CDS
```
Comment (proullon) [inline sdk/action.go]: Strange name for a func creating an Action hehe

File: sdk/cli/cds/action/add.go [hunk 14]
```diff
@@ -12,29 +13,42 @@ import (
...
 }{}
 
 func cmdActionAdd() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "add",
-		Short: "cds action add <actionName>",
-		Long:  ``,
-		Run:   addAction,
+		Short: "cds action add [<actionName>]",
+		Long: `
```
Comment (proullon) [inline sdk/cli/cds/action/add.go]: first line return will be printed

File: engine/api/worker/model.go (hunk 12 above)
Comment (yesnault) [inline engine/api/worker/model.go]: It's a fix, waiting for worker model right refactor. For this case, we have no user, but a group owner.

File: sdk/action.go (hunk 13 above)
Comment (yesnault) [inline sdk/action.go]: Yeah, you're right. This name is same than handler's name on api, and this handler creates an action after "loading" a file (url or file). Now, this handler can insert or update an existing action, with a json data POST by sdk, this is simpliest than post a multipart form, not common in sdk. So name is not great anymore. You may have a new name proposition ? upsertAction ?

File: engine/api/worker/model.go (hunk 12 above)
Comment (proullon) [inline engine/api/worker/model.go]: oh right group owner nevermind :)

File: sdk/action.go (hunk 13 above)
Comment (proullon) [inline sdk/action.go]: hum, CreateFromFile or better ImportAction ?

Comment (fsamin) [inline sdk/action.go]: ImportAction is fine

**Pull request merged**